tools/auth/hubspot.go

```diff
@@ -0,0 +1,92 @@
+package auth
+
+import (
+	"encoding/json"
+	"errors"
+	"time"
+
+	"golang.org/x/oauth2"
+)
+
+var _ Provider = (*Hubspot)(nil)
+
+// NameHubspot is the unique name of the Hubspot provider.
+const NameHubspot string = "hubspot"
+
+// Hubspot allows authentication via Hubspot OAuth2.
+type Hubspot struct {
+	*baseProvider
+}
+
+// NewHubspotProvider creates a new Hubspot provider instance with some defaults.
+func NewHubspotProvider() *Hubspot {
+	return &Hubspot{&baseProvider{
+		scopes: []string{
+			"crm.objects.owners.read",
+		},
+		authUrl:    "https://app.hubspot.com/oauth/authorize",
+		tokenUrl:   "https://api.hubapi.com/oauth/v1/token",
+		userApiUrl: "https://api.hubapi.com/crm/v3/owners",
+	}}
+}
+
+// FetchAuthUser returns an AuthUser instance based on the Hubspot user API.
+// https://developers.hubspot.com/docs/api/crm/owners
+func (p *Hubspot) FetchAuthUser(token *oauth2.Token) (*AuthUser, error) {
+	data, err := p.FetchRawUserData(token)
+	if err != nil {
+		return nil, err
+	}
+
+	rawUser := map[string]any{}
+	if err := json.Unmarshal(data, &rawUser); err != nil {
+		return nil, err
+	}
+
+	var resp struct {
+		Results []struct {
+			ID        string    `json:"id"`
+			Email     string    `json:"email"`
+			FirstName string    `json:"firstName"`
+			LastName  string    `json:"lastName"`
+			UserID    int       `json:"userId"`
+			CreatedAt time.Time `json:"createdAt"`
+			UpdatedAt time.Time `json:"updatedAt"`
+			Archived  bool      `json:"archived"`
+			Teams     []struct {
+				ID      string `json:"id"`
+				Name    string `json:"name"`
+				Primary bool   `json:"primary"`
+			} `json:"teams"`
+		} `json:"results"`
+		Paging struct {
+			Next struct {
+				After string `json:"after"`
+				Link  string `json:"link"`
+			} `json:"next"`
+		} `json:"paging"`
```

Comment:
We don't need to map every field from the response. Map only those fields that are directly used in the `*AuthUser` struct (`id`, `userId`, `email`, `firstName`, `lastName`).

_Side-note: What is the difference between `id` and `userId`?_

Reply:
> Side-note: What is the difference between id and userId?

In Hubspot, users and owners are separate resources, but an owner is also a user. (Weird design)
Here `id` is the owner id since it is sourced from the owner endpoint

From what I know, assigning users to a deal, a contact or a company, the owner id is used
https://community.hubspot.com/t5/APIs-Integrations/User-id-vs-owner-id/m-p/231212

Reply:
Hm, I don't have experience with HubSpot but maybe the `owners` endpoint is not the correct one here. From the docs: 

> HubSpot uses owners to assign CRM objects to specific people in your organization. The endpoints described here are used to get a list of the owners that are available for an account.

Which as far as I understand, more than 1 users/owners could be returned. How we decide which to assign? At the moment we always will extract the first one.

Additionally, maybe we should stick to only the `"oauth"` scope from https://developers.hubspot.com/docs/api/working-with-oauth#scopes.

Reply:
You make good points, so I searched the internet for an answer
I think I've found something [here](https://community.hubspot.com/t5/APIs-Integrations/Get-current-user-id-and-info-from-oAuth-connection/m-p/240602)

Now the only problem is that the URL has a parameter 😅
https://legacydocs.hubspot.com/docs/methods/oauth2/get-access-token-information

Reply:
For the parameter in `GET /oauth/v1/access-tokens/{token}` we can specify a custom `FetchRawUserData` function for the HubSpot provider, something like:

```go
func (p *Hubspot) FetchRawUserData(token *oauth2.Token) ([]byte, error) {
    req, err := http.NewRequest("GET", p.userApiUrl + token.AccessToken, nil)
    if err != nil {
        return nil, err
    }

    return p.sendRawUserDataRequest(req, token)
}
```